### 拼写检查器原理 ###
在所有正确的拼写词中, 我们想要找一个正确的词 c, 使得对于 w 的条件概率最大。求解：  
P(c|w) ->  P(w|c) P(c) / P(w)   
比如：appla是条件w，apple和apply是正确的词c，对于apple和apply来说P(w)都是一样的，所以我们在上式中忽略它, 写成:    
P(w|c) P(c)
* P(c), 文章中出现这个正确拼写的词 c 的概率, 也就是说, 在英语文章中, c 出现的概率有多大。  
假设可以认为单词在文章中出现的概率越大，则正确拼写的概率就越大，可以用单词出现次数来代替这个量。好比说, 英语中出现 the 的概率  P('the') 就相对高, 而出现  P('zxzxzxzyy') 的概率接近0(假设后者也是一个词的话).
* P(w|c), 在用户想键入 c 的情况下敲成 w 的概率。这个是代表用户会以多大的概率把 c 敲错成 w。  

In [1]:
import re

In [2]:
# 读取内容
text = open('big.txt').read()

In [3]:
# 转小写，只保留a-z字符
text = re.findall('[a-z]+', text.lower())

In [4]:
dic_words = {}
for t in text:
    dic_words[t] = dic_words.get(t,0) + 1

In [5]:
dic_words

{'submitted': 27,
 'thiers': 13,
 'communistic': 4,
 'municipality': 4,
 'annals': 5,
 'realizes': 1,
 'coachmen': 6,
 'tantalizing': 1,
 'all': 4144,
 'twot': 1,
 'charcot': 19,
 'vicar': 2,
 'unworthy': 13,
 'begrudge': 1,
 'predominates': 3,
 'consumption': 4,
 'inmate': 1,
 'dolefully': 1,
 'sportsman': 6,
 'episcopalian': 2,
 'assault': 13,
 'befalls': 1,
 'papillomas': 2,
 'fashion': 49,
 'trustfulness': 1,
 'toils': 4,
 'longtemps': 1,
 'colostomy': 1,
 'constructionists': 1,
 'overstretching': 2,
 'clumsy': 8,
 'passport': 4,
 'gauzy': 1,
 'furtively': 2,
 'verrons': 1,
 'contused': 16,
 'conducts': 1,
 'viceroy': 1,
 'avoiding': 22,
 'locus': 1,
 'dorogomilov': 9,
 'nondescript': 1,
 'shortened': 6,
 'carefree': 2,
 'caskets': 1,
 'bronchitis': 1,
 'calculated': 19,
 'rate': 67,
 'save': 110,
 'smacking': 2,
 'predispose': 2,
 'frontiersman': 2,
 'comment': 10,
 'haemorrhages': 14,
 'discover': 28,
 'gazed': 94,
 'administrative': 12,
 'momburg': 1,
 'trestle': 1,
 'hereditary

###  编辑距离: ###
两个词之间的编辑距离定义为使用了几次插入(在词中插入一个单字母), 删除(删除一个单字母), 交换(交换相邻两个字母), 替换(把一个字母换成另一个)的操作从一个词变到另一个词.

In [6]:
# 字母表
alphabet = 'abcdefghijklmnopqrstuvwxyz'

#返回所有与单词 word 编辑距离为 1 的集合
def edits1(word):
    n = len(word)
    return set([word[0:i]+word[i+1:] for i in range(n)] +                     # deletion
               [word[0:i]+word[i+1]+word[i]+word[i+2:] for i in range(n-1)] + # transposition
               [word[0:i]+c+word[i+1:] for i in range(n) for c in alphabet] + # alteration
               [word[0:i]+c+word[i:] for i in range(n+1) for c in alphabet])  # insertion

In [17]:
apple = 'apple'
apple[0:0] + apple[1:]

'pple'

In [18]:
apple = 'apple'
apple[0:1] + apple[2:]

'aple'

In [19]:
apple = 'apple'
apple[0:2] + apple[3:]

'aple'

In [20]:
apple = 'apple'
apple[0:3] + apple[4:]

'appe'

In [21]:
apple = 'apple'
apple[0:4] + apple[5:]

'appl'

In [7]:
#返回所有与单词 word 编辑距离为 2 的集合
#在这些编辑距离小于2的词中间, 只把那些正确的词作为候选词
def edits2(word):
    return set(e2 for e1 in edits1(word) for e2 in edits1(e1))

In [25]:
e1 = edits1('something')
e2 = edits2('something')
len(e1) + len(e2)

114818

与 something 编辑距离为1或者2的单词居然达到了 114,818 个  
优化:只把那些正确的词作为候选词,优化之后edits2只能返回 3 个单词: ‘smoothing’, ‘something’ 和 ‘soothing’

P(w|c)求解：正常来说把一个元音拼成另一个的概率要大于辅音 (因为人常常把 hello 打成 hallo 这样); 把单词的第一个字母拼错的概率会相对小, 等等。但是为了简单起见, 选择了一个简单的方法: 编辑距离为1的正确单词比编辑距离为2的优先级高, 而编辑距离为0的正确单词优先级比编辑距离为1的高.一般把hello打成hallo的可能性比把hello打成halo的可能性大。 

In [51]:
def known(words):
    w = set()
    for word in words:
        if word in dic_words:
            w.add(word)
    return w

# 先计算编辑距离，再根据编辑距离找到最匹配的单词
def correct(word):
    # 获取候选单词
    #如果known(set)非空, candidates 就会选取这个集合, 而不继续计算后面的
    candidates = known([word]) or known(edits1(word)) or known(edits2(word)) or word
    # 字典中不存在相近的词
    if word == candidates:
        return word
    # 返回频率最高的词
    max_num = 0
    for c in candcidates:
        if dic_words[c] >= max_num:
            max_num = dic_words[c]
            candidate = c
    return candidate

In [52]:
#appl #appla #learw #tess #morw
correct('smoothig')

'smoothing'

In [53]:
correct('battl')

'battle'

In [54]:
correct('learww')

'learn'

In [55]:
correct('dagsgasdfeg')

'dagsgasdfeg'